File: Templates/MessageAroundRing/SampleSolutions/Fortran/ringsend.f90
Program RingSend

  use mpi

  Implicit none

  Integer :: world_size, world_rank
  Integer :: up_neigh, dn_neigh
  Integer :: up_send, dn_send
  Integer :: up_sum, dn_sum
  Integer :: my_world

  Integer :: i_round

  Call setup_mpi(my_world, world_size, world_rank)

  print *,"I'm", world_rank, " out of", world_size 

  up_neigh = mod(world_rank+1, world_size)
  dn_neigh = mod(world_rank + world_size -1 , world_size)

  print *,"I'm", world_rank, " my neighbours are", dn_neigh, up_neigh

  up_send = 1000 * world_rank
  dn_send = world_rank

  up_sum = 0
  dn_sum = 0

  Do i_round = 1, world_size
     
     Call exchange(my_world, up_send, dn_send, up_neigh, dn_neigh)
     up_sum=up_sum + up_send
     dn_sum=dn_sum + dn_send
  Enddo

  print *, "I'm", world_rank, " and accumulated", up_sum, dn_sum

  Call shutdown_mpi()

contains

  Subroutine exchange(comm, up_send, dn_send, up_neigh, dn_neigh)
    integer, intent(in) :: comm
    integer, intent(inout) :: up_send, dn_send
    integer, intent(in) :: up_neigh, dn_neigh

    integer :: dn_recv, up_recv
    integer :: merror

    integer :: dn_req, up_req

    Call MPI_Irecv(dn_recv, 1, MPI_INTEGER, dn_neigh, 1, comm, dn_req, merror)
    Call MPI_Irecv(up_recv, 1, MPI_INTEGER, up_neigh, 2, comm, up_req, merror)

    Call MPI_Send(up_send, 1, MPI_INTEGER, up_neigh, 1, comm, merror)
    Call MPI_Send(dn_send, 1, MPI_INTEGER, dn_neigh, 2, comm, merror)

    Call MPI_Wait(dn_req, MPI_STATUS_IGNORE, merror)
    Call MPI_Wait(up_req, MPI_STATUS_IGNORE, merror)

    up_send = dn_recv
    dn_send = up_recv

  end Subroutine exchange

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

  Subroutine setup_mpi(Comm, size, rank)

    integer, intent(out) :: comm
    integer, intent(out) :: size, rank
    Integer :: merror

    Call MPI_Init(merror)

    comm=MPI_COMM_WORLD

    Call MPI_Comm_size(comm, size, merror)
    Call MPI_Comm_rank(comm, rank, merror)

  End Subroutine setup_mpi

  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

  Subroutine shutdown_mpi()

    integer :: merror

    Call MPI_Finalize(merror)

  end Subroutine shutdown_mpi

End Program RingSend

File: Templates/Day_2/MessageAroundRing/SampleSolutions/C/ringsend.c
#include <stdio.h>
#include "mpi.h"


void exchange(MPI_Comm comm, int* pup_send, int* pdn_send,
              int up_neigh, int dn_neigh)
{
    int dn_recv, up_recv;
    MPI_Request dn_req, up_req;
    MPI_Status stat;
    
    MPI_Irecv(&dn_recv, 1, MPI_INT, dn_neigh, 1, comm, &dn_req);
    MPI_Irecv(&up_recv, 1, MPI_INT, up_neigh, 2, comm, &up_req);
    
    MPI_Send(pup_send, 1, MPI_INT, up_neigh, 1, comm);
    MPI_Send(pdn_send, 1, MPI_INT, dn_neigh, 2, comm);
    
    MPI_Wait(&dn_req, &stat);
    MPI_Wait(&up_req, &stat);
    
    *pup_send = dn_recv;
    *pdn_send = up_recv;
    
    return;
}


void setup_mpi(MPI_Comm* pcomm, int* psize, int* prank)
{
    MPI_Init(NULL, NULL);
    
    *pcomm = MPI_COMM_WORLD;
    
    MPI_Comm_size(*pcomm, psize);
    MPI_Comm_rank(*pcomm, prank);
    
    return;
}

void shutdown_mpi()
{
    MPI_Finalize();
    
    return;
}


int main(int argc, char **argv)
{
    int world_size, world_rank;
    int up_neigh, dn_neigh;
   
    int up_send, dn_send;
    int up_sum, dn_sum;
    
    MPI_Comm my_world;
    
    setup_mpi(&my_world, &world_size, &world_rank);
    
    printf("I am rank %i out of %i!\n", world_rank, world_size);
    
    up_neigh = (world_rank + 1)              % world_size;
    dn_neigh = (world_rank + world_size - 1) % world_size;
    
    printf("I'm %i my neighbours are %i and %i\n", world_rank, dn_neigh, up_neigh);
    
    up_send = 1000 * world_rank;
    dn_send = world_rank;
    
    up_sum = 0;
    dn_sum = 0;
    
    for (int i = 0; i < world_size; i++ )
    {
        exchange(my_world, &up_send, &dn_send, up_neigh, dn_neigh);
        up_sum += up_send;
        dn_sum += dn_send;
    }
    
    printf("I am %i and accumulated %i and %i\n", world_rank, dn_sum, up_sum);
    
    shutdown_mpi();
}
